### PR TITLE
Enable compiler warnings in xpack core plugin project

### PR DESCRIPTION
**Pull request by vamuzumd**
This commit is part of issue #[40366](https://github.com/elastic/elasticsearch/issues/40366) to remove disabled lint warnings from gradle files. This PR renables Xlint warnings for XPack/plugin/core project by:-

- Adding wildcard type parameters
- Suppressing unchecked warnings for most occurrences as they involved converting Object into Collection(List, Map) within lambda function.
- Removed dead url -// https://github.com/elastic/x-plugins/issues/724
- Suppressing rawtype warnings for ExecutableChainTransform class as it returns derived class object having more than 1 derived types.
- Adding unchecked suppressions for randomFrom method usages as it returns a generic object.

### CODE REVIEW

File: client/rest-high-level/src/main/java/org/elasticsearch/client/common/XContentSource.java [hunk 1]
```diff
@@ -77,9 +77,8 @@ public boolean isList() {
      * @param path a dot notation path to the requested value
      * @return The extracted value or {@code null} if no value is associated with the given path
      */
-    @SuppressWarnings("unchecked")
-    public <T> T getValue(String path) {
-        return (T) ObjectPath.eval(path, data);
+    public Object getValue(String path) {
```
Comment (vamuzumd) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/common/XContentSource.java:80]: revert

File: libs/x-content/src/main/java/org/elasticsearch/common/xcontent/ObjectPath.java [hunk 2]
```diff
@@ -37,9 +37,8 @@ private ObjectPath() {
      * Return the value within a given object at the specified path, or
      * {@code null} if the path does not exist
      */
-    @SuppressWarnings("unchecked")
-    public static <T> T eval(String path, Object object) {
-        return (T) evalContext(path, object);
+    public static Object eval(String path, Object object) {
```
Comment (vamuzumd) [inline libs/x-content/src/main/java/org/elasticsearch/common/xcontent/ObjectPath.java:40]: revert

**Comment by vamuzumd** [comment]
Hi @martijnvg , @pugnascotia , could you please review this PR?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/watcher/transform/chain/ExecutableChainTransform.java [hunk 3]
```diff
@@ -24,22 +24,27 @@
 
 public class ExecutableChainTransform extends ExecutableTransform<ChainTransform, ChainTransform.Result> {
 
+    @SuppressWarnings("rawtypes")
```
Comment (pugnascotia) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/watcher/transform/chain/ExecutableChainTransform.java:27]: I think we should just put this annotation on the class.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/watcher/support/xcontent/XContentSource.java [hunk 4]
```diff
@@ -76,8 +77,8 @@ public boolean isMap() {
     /**
      * @return The source as a map
      */
-    public Map<String, Object> getAsMap() {
-        return (Map<String, Object>) data();
+    public Map<?, ?> getAsMap() {
```
Comment (pugnascotia) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/watcher/support/xcontent/XContentSource.java:80]: I understand why you've made this change, but actually I think we want to keep this code the way it was. Can you revert this change please?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/watcher/support/xcontent/XContentSource.java [hunk 5]
```diff
@@ -90,8 +91,7 @@ public boolean isList() {
     /**
      * @return The source as a list
      */
-    public List<Object> getAsList() {
-        return (List<Object>) data();
+    public List<?> getAsList() { return ((List<?>) data());
```
Comment (pugnascotia) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/watcher/support/xcontent/XContentSource.java:94]: Same here, can you revert this change please?

**Review by pugnascotia — CHANGES_REQUESTED** [review]
Thank you very much @vamuzumd, the PR needs a couple of tweaks but otherwise looks good.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/watcher/support/xcontent/XContentSource.java (hunk 5 above)
Comment (vamuzumd) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/watcher/support/xcontent/XContentSource.java]: Reverted the changes in this file except one suppressed tag

**Comment by vamuzumd** [comment]
> Thank you very much @vamuzumd, the PR needs a couple of tweaks but otherwise looks good.

Thanks for the feedback @pugnascotia !

**Review by breskeby — APPROVED** [review]
lgtm apart from what @pugnascotia already mentioned

**Comment by pugnascotia** [comment]
@elasticmachine test this please

**Comment by pugnascotia** [comment]
@vamuzumd there are compilation failures. I suggest running `./gradlew precommit` to make sure the code all compiles and passes our style checks.

**Comment by pugnascotia** [comment]
Hi @vamuzumd, are you still interested in carrying this PR forwards?

**Comment by vamuzumd** [comment]
> Hi @vamuzumd, are you still interested in carrying this PR forwards?

Sorry for the late, I will work towards this PR and send an update soon.

**Comment by vamuzumd** [comment]
> @vamuzumd there are compilation failures. I suggest running `./gradlew precommit` to make sure the code all compiles and passes our style checks.

@pugnascotia, thanks! I just resolved all compilation and code check failures and pushed new changes, can you take a look?

**Comment by pugnascotia** [comment]
@vamuzumd would you mind fetching and merging `master`? There are some conflicts to resolve.

**Comment by vamuzumd** [comment]
> @vamuzumd would you mind fetching and merging `master`? There are some conflicts to resolve.

whoops, thanks for pointing out! Just resolved the conflicts.

**Comment by vamuzumd** [comment]
> > @vamuzumd would you mind fetching and merging `master`? There are some conflicts to resolve.
> 
> whoops, thanks for pointing out! Just resolved the conflicts. Working on build failures in  the pipeline.

**Comment by vamuzumd** [comment]
> > > @vamuzumd would you mind fetching and merging `master`? There are some conflicts to resolve.
> > 
> > 
> > whoops, thanks for pointing out! Just resolved the conflicts. Working on build failures in  the pipeline.

Also resolve build failures. the PR looks ready for merge.

**Comment by vamuzumd** [comment]
> Thank you very much @vamuzumd, the PR needs a couple of tweaks but otherwise looks good.

Thanks, @pugnascotia , I've resolved build issues and merge conflicts. Please lmk if the last iteration looks good.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/config/RuleScope.java [hunk 6]
```diff
@@ -43,7 +44,7 @@
             Map<String, FilterRef> scope = new HashMap<>();
             for (Map.Entry<String, Object> entry : unparsedScope.entrySet()) {
                 try (XContentBuilder builder = XContentFactory.jsonBuilder()) {
-                    builder.map((Map<String, ?>) entry.getValue());
+                    builder.map((Map<String, Object>) entry.getValue());
```
Comment (pugnascotia) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/config/RuleScope.java:47]: `builder.map` expects a `Map<String, ?>`, so I don't think we need to change this line.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/watcher/actions/ActionWrapper.java [hunk 7]
```diff
@@ -174,7 +175,7 @@ public ActionWrapperResult execute(WatchExecutionContext ctx) {
                 Object object = ObjectPath.eval(path, toMap(ctx));
                 int runs = 0;
                 if (object instanceof Collection) {
-                    Collection collection = Collection.class.cast(object);
+                    Collection<?> collection = Collection.class.cast(object);
```
Comment (pugnascotia) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/watcher/actions/ActionWrapper.java:178]: I think in this case, we can just say:
```suggestion
                    Collection<?> collection = (Collection<?>) object;
```

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/action/TransportXPackInfoActionTests.java [hunk 8]
```diff
@@ -43,6 +43,7 @@
 
 public class TransportXPackInfoActionTests extends ESTestCase {
 
+    @SuppressWarnings("unchecked")
```
Comment (pugnascotia) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/action/TransportXPackInfoActionTests.java:46]: The other option here, rather than suppressing the warning for the whole method, is to define a helper like so:

```java
    @SuppressWarnings("unchecked")
    private <T> ActionListener<T> anyActionListener() {
        return any(ActionListener.class);
    }
```

then replace `any(ActionListener.class)` with `anyActionListener()`. It's slightly annoying to have to do this, but it's an easy way to limit the scope of the suppressions.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/ClassificationTests.java [hunk 9]
```diff
@@ -89,6 +89,7 @@ protected NamedWriteableRegistry getNamedWriteableRegistry() {
         return new NamedWriteableRegistry(entries);
     }
 
+    @SuppressWarnings({"unchecked"})
```
Comment (pugnascotia) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/ClassificationTests.java:92]: I don't believe we need this here?

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/RegressionTests.java [hunk 10]
```diff
@@ -82,6 +82,7 @@ public static Regression createRandom() {
         return createRandom(BoostedTreeParamsTests.createRandom());
     }
 
+    @SuppressWarnings({"unchecked"})
```
Comment (pugnascotia) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/RegressionTests.java:85]: I don't believe we need this here?

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/inference/InferenceToXContentCompressorTests.java [hunk 11]
```diff
@@ -22,6 +22,7 @@
 
 import static org.hamcrest.Matchers.equalTo;
 
+@SuppressWarnings({"unchecked"})
```
Comment (pugnascotia) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/inference/InferenceToXContentCompressorTests.java:25]: I don't believe we need this here? That's what my IDE is telling me?

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/inference/NamedXContentObjectsTests.java [hunk 12]
```diff
@@ -147,6 +147,7 @@ public void chooseStrictOrLenient() {
         lenient = randomBoolean();
     }
 
+    @SuppressWarnings({"unchecked"})
```
Comment (pugnascotia) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/inference/NamedXContentObjectsTests.java:150]: This too, do we actually need this?

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/ClassificationTests.java (hunk 9 above)
Comment (pugnascotia) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/ClassificationTests.java:92]: Turns out my IDE was lying - ignore me!

Comment (pugnascotia) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/ClassificationTests.java:92]: Ah, I see what the root problem is here. What we actually want to do is edit `ESTestCase.randomFrom` as follows. Then you won't need to add `@SuppressWarnings({"unchecked"})` to test methods that call `randomFrom()`.

```diff
diff --git a/test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java b/test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java
index a18fd0d6bcd..991550fe334 100644
--- a/test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java
+++ b/test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java
@@ -713,6 +713,8 @@ public abstract class ESTestCase extends LuceneTestCase {
     }
 
     /** Pick a random object from the given array. The array must not be empty. */
+    @SafeVarargs
+    @SuppressWarnings("varargs")
     public static <T> T randomFrom(T... array) {
         return randomFrom(random(), array);
     }
```

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/utils/MlIndexAndAliasTests.java [hunk 13]
```diff
@@ -81,6 +81,7 @@
     private ArgumentCaptor<CreateIndexRequest> createRequestCaptor;
     private ArgumentCaptor<IndicesAliasesRequest> aliasesRequestCaptor;
 
+    @SuppressWarnings("unchecked")
```
Comment (pugnascotia) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/utils/MlIndexAndAliasTests.java:84]: We can use the `anyActionListener()` trick here again, as well as defining a `mockActionListener()` for the other compiler error.

**Comment by pugnascotia** [comment]
@elasticmachine ok to test

**Comment by pugnascotia** [comment]
@vamuzumd this is looking good, just a few things to fix up. I appreciate you tackling this work, some of the generics are...interesting 😱

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/action/TransportXPackInfoActionTests.java (hunk 8 above)
Comment (vamuzumd) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/action/TransportXPackInfoActionTests.java:46]: That's a great suggestion! Unfortunately, that would still require adding suppression since the method does castings like  (ActionListener<XPackInfoFeatureResponse>)answer.getArguments()[2] on line 58. So, keeping this as is.

Comment (vamuzumd) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/action/TransportXPackInfoActionTests.java:46]: Plus it would require adding helpers for eq on line 57 and any(ActionRequest) which could get a bit gawky, but happy to change this if you feel otherwise.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/utils/MlIndexAndAliasTests.java (hunk 13 above)
Comment (vamuzumd) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/utils/MlIndexAndAliasTests.java:84]: Sure, we could do that, but like the other case, this requires defining 4 helpers plus address the casting on line 100, hence leaving this as is. But please lmk if you think otherwise.

**Comment by vamuzumd** [comment]
> @vamuzumd this is looking good, just a few things to fix up. I appreciate you tackling this work, some of the generics are...interesting 😱

Thanks @pugnascotia, I've addressed your feedback.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/ClassificationTests.java (hunk 9 above)
Comment (vamuzumd) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/ClassificationTests.java]: Grat suggestion! Significantly reduced number of files changed.

**Comment by vamuzumd** [comment]
> @vamuzumd there are compilation failures. I suggest running `./gradlew precommit` to make sure the code all compiles and passes our style checks.



> > @vamuzumd this is looking good, just a few things to fix up. I appreciate you tackling this work, some of the generics are...interesting 😱
> 
> Thanks @pugnascotia, I've addressed your feedback.

@pugnascotia , the ci/bwc build fails for 2 tests in org.elasticsearch.backwards.MixedClusterClientYamlTestSuiteIT suite.
I don't see these failures associated with my changes as they belong to different project/package from XPack. Any suggestions?

**Comment by vamuzumd** [comment]
Seems like the test BootStrapTests.testTriggeredWatchLoading failed for elasticsearch-ci/2, which seems flaky and unrelated to my change based on error. @pugnascotia, thoughts?

Partial stacktrace

java.lang.AssertionError: Count is 18 hits but 9 was expected.  Total shards: 1 Successful shards: 1 & 0 shard failures:
	at __randomizedtesting.SeedInfo.seed([76DF43B200299F27:4F3123083FC3A7D8]:0)
	at org.junit.Assert.fail(Assert.java:88)
	at org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertHitCount(ElasticsearchAssertions.java:259)
	at org.elasticsearch.xpack.watcher.test.integration.BootStrapTests.lambda$assertSingleExecutionAndCompleteWatchHistory$3(BootStrapTests.java:289)

**Comment by vamuzumd** [comment]
@pugnascotia, could you please review/sign-off this PR? Just hoping to avoid another merge conflict since the PR has a large number of files changed.

**Comment by pugnascotia** [comment]
Sorry @vamuzumd, I've been on vacation for the past week.

@elasticmachine update branch

**Comment by pugnascotia** [comment]
@elasticmachine update branch

**Review by pugnascotia — APPROVED** [review]
LGTM!

**Comment by pugnascotia** [comment]
Thank you for your work @vamuzumd!

**Comment by pugnascotia** [comment]
@vamuzumd sorry for the noise, I should have added the right labels sooner.

**Comment by pugnascotia** [comment]
Backported to `7.x` in 37a24d58721
